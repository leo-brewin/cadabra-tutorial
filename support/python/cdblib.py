def create (file_name):
  import json, io, os, errno

  try:
      os.remove(file_name)                # delete the file if it exsists
      with open(file_name, 'w'): pass     # create an empty file
  except OSError as e:
      if e.errno == errno.ENOENT:         # errno.ENOENT = no such file or directory
         with open(file_name, 'w'): pass  # create an empty file
      else:
          raise                           # report an exception

  # Create and save an empty dict
  data_out = {}
  with io.open(os.getcwd() + '/' + file_name, 'w', encoding='utf-8') as out_file:
      out_file.write(json.dumps(data_out,
                                indent=2,
                                sort_keys=True,
                                separators=(',', ': '),
                                ensure_ascii=False)+'\n')

# put and get based on ~/Python/examples/dict-io/lcb-01.py

def put (key_name,object,file_name):
  import json, io, os
  import cadabra2
  __cdbkernel__ = cadabra2.__cdbkernel__

  # Read the current dict
  with io.open(os.getcwd() + '/' + file_name) as inp_file:
      data_out = json.load(inp_file)

  # Add a new entry to the dict.

  # Two choices here. Both give almost identical output except for the rule symbol '->'.
  # Using str(object) will write this using unicode characters. This causes problems
  # down the track -- when these rules are imported in other codes we get a Python
  # error -- it reports that it can't handle this unicode character.
  # Using object.input_form() does the trick. It writes out '->'. This is good.

  # data_out[key_name] = str(object)       # will write rule char using unicode
  data_out[key_name] = object.input_form() # will write rule char as '->'

  # Save the updated dict
  with io.open(os.getcwd() + '/' + file_name, 'w', encoding='utf-8') as out_file:
      out_file.write(json.dumps(data_out,
                                indent=2,
                                sort_keys=True,
                                separators=(',', ': '),
                                ensure_ascii=False)+'\n')

def get (key_name,file_name):
  import json, io, os
  import cadabra2
  __cdbkernel__ = cadabra2.__cdbkernel__

  # Read the current dict
  with io.open(os.getcwd() + '/' + file_name) as inp_file:
      data_inp = json.load(inp_file)

  # Return one entry from the dict
  return cadabra2.Ex (data_inp[key_name])
