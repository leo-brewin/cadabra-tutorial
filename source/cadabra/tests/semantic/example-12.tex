\documentclass[12pt]{cdblatex}
\usepackage{fancyhdr}
\usepackage{footer}

\begin{document}

% ============================================================================================
\section*{Example 12 Checking the 2nd and 3rd order terms of Calzetta etal.}

\CdbSetup{action=hide}

\begin{cadabra}
   {a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,u,v#}::Indices("latin",position=independent).
   {\mu,\nu,\rho,\sigma,\tau,\lambda,\xi#}::Indices("greek",position=independent).

   \nabla{#}::Derivative.

   g_{a b}::Metric.
   g^{a b}::InverseMetric.

   \delta{#}::KroneckerDelta.

   R_{a b c d}::RiemannTensor.

   def check (objA,objB):

       tmp := @(objA) - @(objB).

       distribute     (tmp)
       sort_product   (tmp)
       rename_dummies (tmp)
       canonicalise   (tmp)

       return tmp

   import cdblib

   # ---------------------------------------------
   check000A = cdblib.get ('check000','expected/example-12.json')
   check001A = cdblib.get ('check001','expected/example-12.json')
   check002A = cdblib.get ('check002','expected/example-12.json')
   check003A = cdblib.get ('check003','expected/example-12.json')

   # ---------------------------------------------
   check000B = cdblib.get ('check000','output/example-12.json')
   check001B = cdblib.get ('check001','output/example-12.json')
   check002B = cdblib.get ('check002','output/example-12.json')
   check003B = cdblib.get ('check003','output/example-12.json')

   # ---------------------------------------------
   diff000 = check (check000A,check000B)   # cdb (ex-12.diff000,diff000)
   diff001 = check (check001A,check001B)   # cdb (ex-12.diff001,diff001)
   diff002 = check (check002A,check002B)   # cdb (ex-12.diff002,diff002)
   diff003 = check (check003A,check003B)   # cdb (ex-12.diff003,diff003)

\end{cadabra}

% \clearpage

\begin{dgroup*}
   \Dmath*{\cdb*{ex-12.diff000}}
   \Dmath*{\cdb*{ex-12.diff001}}
   \Dmath*{\cdb*{ex-12.diff002}}
   \Dmath*{\cdb*{ex-12.diff002}}
\end{dgroup*}

\end{document}
