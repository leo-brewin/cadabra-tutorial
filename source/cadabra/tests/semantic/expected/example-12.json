{
  "check000": "z^{h} z^{i} ( 1/3 g^{a b} x^{c} (R_{h b i c} + R_{h c i b}) +  1/12 g^{a b} x^{c} x^{d} (2\\nabla_{c}(R_{h b i d}) + \\nabla_{b}(R_{h c i d})-\\nabla_{i}(R_{h c b d}) + 2\\nabla_{c}(R_{h d i b})-\\nabla_{h}(R_{i c b d})) +  1/40 g^{a b} x^{c} x^{d} x^{e} (2\\nabla_{c d}(R_{h b i e}) + \\nabla_{b c}(R_{h d i e}) + \\nabla_{c b}(R_{h d i e})-\\nabla_{i c}(R_{h d b e})-\\nabla_{c i}(R_{h d b e}) + 2\\nabla_{c d}(R_{h e i b})-\\nabla_{h c}(R_{i d b e})-\\nabla_{c h}(R_{i d b e})) +  1/45 g^{a b} g^{c d} x^{e} x^{f} x^{g} (-2R_{h b c e} R_{i f d g} + 4R_{h c i e} R_{b f d g}-R_{h c b e} R_{i f d g} + 4R_{h e i c} R_{b f d g} + R_{h e b c} R_{i f d g}-2R_{h e c f} R_{i b d g}-R_{h e c f} R_{i d b g} + R_{h e c f} R_{i g b d}))",
  "check001": " 2/3 g^{a b} x^{c} z^{d} z^{e} R_{b d c e} +  1/12 g^{a b} x^{c} x^{d} z^{e} z^{f} (4\\nabla_{c}(R_{b e d f}) + 2\\nabla_{e}(R_{b c d f}) + \\nabla_{b}(R_{c e d f})) +  1/40 g^{a b} x^{c} x^{d} x^{e} z^{f} z^{g} (4\\nabla_{c d}(R_{b f e g}) + 2\\nabla_{c f}(R_{b d e g}) + 2\\nabla_{f c}(R_{b d e g}) + \\nabla_{b c}(R_{d f e g}) + \\nabla_{c b}(R_{d f e g})) +  2/45 g^{a b} g^{c d} x^{e} x^{f} x^{g} z^{h} z^{i} (-2R_{b h c e} R_{d f g i}-R_{b e c h} R_{d f g i} + 4R_{b e c f} R_{d h g i} + R_{b c e h} R_{d f g i})",
  "check002": " 2/3 x^{b} z^{c} z^{d} R_{a c b d} +  1/6 x^{b} x^{c} z^{d} z^{e} \\nabla_{d}(R_{a b c e}) +  1/3 x^{b} x^{c} z^{d} z^{e} \\nabla_{b}(R_{a d c e}) +  1/12 x^{b} x^{c} z^{d} z^{e} \\nabla_{a}(R_{b d c e}) +  1/20 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{b e}(R_{a c d f}) +  1/20 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{e b}(R_{a c d f}) +  1/10 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{b c}(R_{a e d f}) +  1/40 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{a b}(R_{c e d f}) +  1/40 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{b a}(R_{c e d f}) +  2/45 g^{b c} x^{d} x^{e} x^{f} z^{g} z^{h} R_{a b d g} R_{c e f h} +  8/45 g^{b c} x^{d} x^{e} x^{f} z^{g} z^{h} R_{a d b e} R_{c g f h} -  2/45 g^{b c} x^{d} x^{e} x^{f} z^{g} z^{h} R_{a d b g} R_{c e f h} -  4/45 g^{b c} x^{d} x^{e} x^{f} z^{g} z^{h} R_{a g b d} R_{c e f h}",
  "check003": " 2/3 x^{b} z^{c} z^{d} R_{a c b d} +  1/12 x^{b} x^{c} z^{d} z^{e} (2\\nabla_{d}(R_{a b c e}) + 4\\nabla_{b}(R_{a d c e}) + \\nabla_{a}(R_{b d c e})) +  1/40 x^{b} x^{c} x^{d} z^{e} z^{f} (2\\nabla_{b e}(R_{a c d f}) + 2\\nabla_{e b}(R_{a c d f}) + 4\\nabla_{b c}(R_{a e d f}) + \\nabla_{a b}(R_{c e d f}) + \\nabla_{b a}(R_{c e d f})) +  2/45 g^{b c} x^{d} x^{e} x^{f} z^{g} z^{h} (R_{a b d g} R_{c e f h} + 4R_{a d b e} R_{c g f h}-R_{a d b g} R_{c e f h}-2R_{a g b d} R_{c e f h})",
  "check004": " -  2/3 x^{b} z^{c} z^{d} R_{a d c b} -  1/12 x^{b} x^{c} z^{d} z^{e} \\nabla_{e}(R_{a c d b}) -  5/12 x^{b} x^{c} z^{d} z^{e} \\nabla_{c}(R_{a e d b}) -  1/40 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{f c}(R_{a d e b}) -  1/40 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{c f}(R_{a d e b}) -  3/20 x^{b} x^{c} x^{d} z^{e} z^{f} \\nabla_{c d}(R_{a f e b}) -  37/360 x^{b} x^{c} x^{d} z^{e} z^{f} R_{a d g b} g^{g h} R_{h e f c} +  2/15 x^{b} x^{c} x^{d} z^{e} z^{f} R_{a g e b} g^{g h} R_{h c f d} +  7/120 x^{b} x^{c} x^{d} z^{e} z^{f} R_{a d g e} g^{g h} R_{h b f c}",
  "check005": " 2/3 x^{b} z^{c} z^{d} R_{a c b d} +  1/12 x^{b} x^{c} z^{d} z^{e} (\\nabla_{d}(R_{a b c e}) + 5\\nabla_{b}(R_{a d c e})) +  1/40 x^{b} x^{c} x^{d} z^{e} z^{f} (\\nabla_{b e}(R_{a c d f}) + \\nabla_{e b}(R_{a c d f}) + 6\\nabla_{b c}(R_{a e d f})) +  1/360 g^{b c} x^{d} x^{e} x^{f} z^{g} z^{h} (37R_{a d b e} R_{c g f h}-21R_{a d b g} R_{c e f h} + 48R_{a b d g} R_{c e f h})"
}
